cou/steps/execute.py

```diff
@@ -25,6 +26,66 @@
 logger = logging.getLogger(__name__)
 
 
+async def _run_sub_stpes_in_parallel(
+    step: BaseStep, prompt: bool, overwrite_progress: bool
+) -> None:
+    """Run all sub-steps of step in parallel.
+
+    :param step: Step to be executed.
+    :type step: BaseStep
+    :param prompt: Whether to run upgrade step with prompt (interactive mode).
+    :type prompt: bool
+    :param overwrite_progress: Whether to overwrite the progress indication message in CLI output
+                               for all sub-steps. True to overwrite and False (the default) to
+                               persist.
+    :type overwrite_progress: bool
+    :raises HaltUpgradeExecution: When any step raised it. All other steps will be completed.
+    """
+    logger.debug("running all sub-steps of %s step in parallel", step)
+    grouped_coroutines = (
+        apply_step(sub_step, prompt, overwrite_progress) for sub_step in step.sub_steps
+    )
+    try:
+        await asyncio.gather(*grouped_coroutines)
+    except HaltUpgradeExecution as error:
+        logger.debug("%s", error)
+        raise
+
+
+async def _run_sub_stpes_sequentially(
+    step: BaseStep, prompt: bool, overwrite_progress: bool
+) -> None:
+    """Run all sub-steps of step sequentially.
+
+    :param step: Step to be executed.
+    :type step: BaseStep
+    :param prompt: Whether to run upgrade step with prompt (interactive mode).
+    :type prompt: bool
+    :param overwrite_progress: Whether to overwrite the progress indication message in CLI output
+                               for all sub-steps. True to overwrite and False (the default) to
+                               persist.
+    :type overwrite_progress: bool
+    :raises HaltUpgradeExecution: When any step raised it. Remaining steps will be skipped.
+    """
+    logger.debug("running all sub-steps of %s step sequentially", step)
+    for sub_step in step.sub_steps:
+        logger.debug("running sub-step %s of %s step", sub_step, step)
+        try:
+            await apply_step(sub_step, prompt, overwrite_progress)
+        except HaltUpgradeExecution:
+            sub_step_index = step.sub_steps.index(sub_step)
+            logger.warning(
+                "Step: '%s' from '%s' failed to complete execution. The following steps will "
+                "be skipped:\n %s",
+                sub_step.description,
+                step.description,
+                "\n".join(
+                    sub_step.description for sub_step in step.sub_steps[sub_step_index + 1 :]
```

Comment:
I know that step dependency is not implemented in `main` yet, but just want to put a reminder here to add that once step implementation is here ([ref](https://github.com/canonical/charmed-openstack-upgrader/pull/239/commits/611ff2838aa03a28928ab2805d3163717fbdf9a0#diff-391beb5540d6d993d384de19d144db9956db73d354d26a99ceef4a813fa42f33R70)).

Reply:
The dependent is already introduced in main, but here it does not make sense to use it, since the current logic is skipping any further steps and not only the dependent ones. So the logs "lied".

Reply:
As I mentioned in the original PR, the problem with not using dependency is because including `Enable nova-compute scheduler from unit: 'nova-compute/2' step` in the skipped step gives the user a false impression that the scheduler is not re-enabled upon exiting.

This is not an ideal solution either because I think the correct way is to have `re-enabling scheduler` as a clean-up step (an actual step, not running inside check empty function) to always run. But with the log change here, at least we are not misleading the user.